rat/ops.py

```diff
@@ -60,14 +107,87 @@ def get_key(self):
     def code(self):
         return f"subscripts['{self.variable.code()}']"
 
+    def __post_init__(self):
+        self.out_type = types.SubscriptType
+
+    def fold(self):
+        return self
+
+    def __str__(self):
+        return f"Subscript(names=({', '.join(x.__str__() for x in self.names)}), shift=({', '.join(x.__str__() for x in self.shifts)}))"
+
+
+@dataclass
+class SubscriptOp(Expr):
+    """
+    SubscriptOp = operator[] (Subscript, ..., Subscript)
+    (Subscript, ..., Subscript) -> Subscript
+    """
+
+    subscripts: List[Expr]
+
+    def __post_init__(self):
+        signatures = {(types.SubscriptType,) * len(self.subscripts): types.SubscriptType}
+        self.out_type = types.get_output_type(signatures, tuple([x.out_type for x in self.subscripts]))
+
+    def fold(self) -> Subscript:
+        """
+        Combines multiple subscripts into a single subscript
+        """
+        names: List[str] = []
+        shifts: List[Union[Expr, None]] = []
+        for subscript in self.subscripts:
+            subscript = subscript.fold()
+            shifts.extend(subscript.shifts)
+            names.extend(subscript.names)
+
+        return Subscript(names=tuple(names), shifts=tuple(shifts), line_index=self.line_index, column_index=self.column_index)
```

Comment:
Oh interesting so what is happening here is for an expression like:

```
variable[A, B, C]
```

This is parsed as:
```
variable[subscript_A, subscript_B, subscript_C]
```

And then we fold things together to be:
```
variable[subscript_A_B_C]
```

This about right?

Reply:
Actually I see the folding happens through basically everything -- is the folding done exclusively to bundle the subscripts together? Or are there other types of constants?

Reply:
> Oh interesting so what is happening here is for an expression like:
> 
> ```
> variable[A, B, C]
> ```
> 
> This is parsed as:
> 
> ```
> variable[subscript_A, subscript_B, subscript_C]
> ```
> 
> And then we fold things together to be:
> 
> ```
> variable[subscript_A_B_C]
> ```
> 
> This about right?

Yes. Same goes for shifts. We bundle multiple subscripts into a single one.

Reply:
> Actually I see the folding happens through basically everything -- is the folding done exclusively to bundle the subscripts together? Or are there other types of constants?

Yup for constants too. If you check the parser unittests you can see some examples.